File: cs/Day11.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace cs
{
    class Day11
    {
        private static void flash(List<List<Tuple<int, bool>>> grid, int x, int y, ref int counter)
        {
            if (grid[y][x].Item2)
                return;

            grid[y][x] = new Tuple<int, bool>(grid[y][x].Item1 + 1, false);

            if (grid[y][x].Item1 > 9)
            {
                grid[y][x] = new Tuple<int, bool>(0, true);
                counter++;

                foreach (int y_off in new int[] { -1, 0, 1 })
                    foreach (int x_off in new int[] { -1, 0, 1 })
                        if (y + y_off >= 0 && y + y_off < grid.Count && x + x_off >= 0 && x + x_off < grid[y + y_off].Count)
                            flash(grid, x + x_off, y + y_off, ref counter);
            }
        }

        public static void Part1()
        {
            List<List<Tuple<int, bool>>> grid = new List<List<Tuple<int, bool>>>();

            File.ReadLines(@"../../../../../data/day11.txt").ToList().ForEach(line =>
            {
                grid.Add(new List<Tuple<int, bool>>());
                line.ToCharArray().ToList().ForEach(num =>
                {
                    grid.Last().Add(new Tuple<int, bool>(num - '0', false));
                });
            });

            int counter = 0;
            for (int step = 0; step < 100; step++)
            {
                for (int row = 0; row < grid.Count; row++)
                    for (int col = 0; col < grid[row].Count; col++)
                        grid[row][col] = new Tuple<int, bool>(grid[row][col].Item1, false);

                for (int row = 0; row < grid.Count; row++)
                    for (int col = 0; col < grid[row].Count; col++)
                        flash(grid, col, row, ref counter);
            }

            Console.WriteLine("Day: 11 | Part: 1 | Result: " + counter);
        }

        public static void Part2()
        {
            List<List<Tuple<int, bool>>> grid = new List<List<Tuple<int, bool>>>();

            File.ReadLines(@"../../../../../data/day11.txt").ToList().ForEach(line =>
            {
                grid.Add(new List<Tuple<int, bool>>());
                line.ToCharArray().ToList().ForEach(num =>
                {
                    grid.Last().Add(new Tuple<int, bool>(num - '0', false));
                });
            });

            int ignore = 0;
            int step = 0;
            for (; grid.All(row => row.All(col => col.Item2)) == false; step++)
            {
                for (int row = 0; row < grid.Count; row++)
                    for (int col = 0; col < grid[row].Count; col++)
                        grid[row][col] = new Tuple<int, bool>(grid[row][col].Item1, false);

                for (int row = 0; row < grid.Count; row++)
                    for (int col = 0; col < grid[row].Count; col++)
                        flash(grid, col, row, ref ignore);
            }

            Console.WriteLine("Day: 11 | Part: 2 | Result: " + step);
        }
    }
}

File: python/Day11.py
def flash(grid: list[list[list[int, bool]]], x: int, y: int, counter: int) -> int:
    if grid[y][x][1]:
        return counter

    grid[y][x][0] += 1

    if grid[y][x][0] > 9:
        grid[y][x][0] = 0
        grid[y][x][1] = True
        counter += 1

        for y_off in [-1, 0, 1]:
            for x_off in [-1, 0, 1]:
                if y + y_off in range(len(grid)) and x + x_off in range(
                    len(grid[y + y_off])
                ):
                    counter = flash(grid, x + x_off, y + y_off, counter)

    return counter


def part_1() -> None:
    grid: list[list[list[int, bool]]] = []

    with open("../data/day11.txt", "r") as dFile:
        for line in dFile.readlines():
            grid.append([])
            for num in line.strip():
                grid[-1].append([int(num), False])

    counter: int = 0
    for _ in range(1, 101):
        for row in grid:
            for col in row:
                col[1] = False

        for row in range(len(grid)):
            for col in range(len(grid[row])):
                counter = flash(grid, col, row, counter)

    print("Day: 11 | Part: 1 | Result:", counter)


def part_2() -> None:
    grid: list[list[list[int, bool]]] = []

    with open("../data/day11.txt", "r") as dFile:
        for line in dFile.readlines():
            grid.append([])
            for num in line.strip():
                grid[-1].append([int(num), False])

    step: int = 0
    while not all(map(lambda row: all(map(lambda col: col[1], row)), grid)):
        step += 1
        for row in grid:
            for col in row:
                col[1] = False

        for row in range(len(grid)):
            for col in range(len(grid[row])):
                flash(grid, col, row, 0)

    print("Day: 11 | Part: 2 | Result:", step)


if __name__ == "__main__":
    part_1()
    part_2()
